src/core/mempool.rs

```diff
@@ -1360,16 +1640,231 @@ impl MemPoolDB {
+    /// (so if some nodes are configured to return fewer than MAX_BLOOM_COUNTER_TXS transactions,
+    /// a requesting node will still have a good chance of getting something useful).
+    pub fn find_next_missing_transactions(
+        &self,
+        data: &MemPoolSyncData,
+        height: u64,
+        last_txid: &Txid,
+        max_txs: u64,
+        max_run: u64,
+    ) -> Result<Vec<StacksTransaction>, db_error> {
+        let mut ret = vec![];
+        let sql = "SELECT mempool.txid as txid, mempool.tx as tx \
+                   FROM mempool JOIN randomized_txids \
+                   ON mempool.txid = randomized_txids.txid \
+                   WHERE randomized_txids.hashed_txid > ?1 \
+                   AND mempool.height > ?2 \
+                   AND NOT EXISTS \
+                        (SELECT 1 FROM removed_txids WHERE txid = mempool.txid) \
+                   ORDER BY randomized_txids.hashed_txid ASC LIMIT ?3";
+
+        let args: &[&dyn ToSql] = &[
+            &last_txid,
+            &u64_to_sql(height.saturating_sub(BLOOM_COUNTER_DEPTH as u64))?,
+            &u64_to_sql(max_run)?,
+        ];
+
+        let mut tags_table = HashSet::new();
+        if let MemPoolSyncData::TxTags(_, ref tags) = data {
+            for tag in tags.iter() {
+                tags_table.insert(tag.clone());
+            }
+        }
+
+        let mut stmt = self.conn().prepare(sql)?;
+        let mut rows = stmt.query(args)?;
+        while let Some(row) = rows.next()? {
+            let txid = Txid::from_column(row, "txid")?;
+            test_debug!("Consider txid {}", &txid);
+            let contains = match data {
+                MemPoolSyncData::BloomFilter(ref bf) => bf.contains_raw(&txid.0),
+                MemPoolSyncData::TxTags(ref seed, ..) => {
+                    tags_table.contains(&TxTag::from_seed_and_txid(seed, &txid))
+                }
+            };
+            if contains {
+                // remote peer already has this one
+                continue;
+            }
+
+            let tx_bytes: Vec<u8> = row.get_unwrap("tx");
+            let tx = StacksTransaction::consensus_deserialize(&mut &tx_bytes[..])
+                .map_err(|_e| db_error::ParseError)?;
+
+            test_debug!("Returning txid {}", &txid);
+            ret.push(tx);
+            if (ret.len() as u64) >= max_txs {
+                break;
+            }
+        }
+
+        Ok(ret)
+    }
+
+    /// Stream transaction data
+    pub fn stream_txs<W: Write>(
+        &self,
+        fd: &mut W,
+        query: &mut TxStreamData,
+        count: u64,
+    ) -> Result<u64, ChainstateError> {
+        let mut num_written = 0;
+        while num_written < count {
+            if query.num_txs >= query.max_txs {
+                // don't serve more than this many txs
+                break;
+            }
+
+            // write out bufferred tx
+            while query.tx_buf.len() > 0 && query.tx_buf_ptr < query.tx_buf.len() {
```

Comment:
It does iterate, but it shouldn't -- the API contract is that "no more than `count` bytes shall be written out."  This code was violating this contract, but was successfully streaming out data.  I didn't notice it because the underlying `NetworkConnection<StacksHttp>` (a.k.a. `ConnectionHttp`) instance uses a `Pipe` instance to buffer writes from serialization code (which it will gladly fill up with a pending `StacksTransaction`'s bytes even if the HTTP server, which owns the `ConversationHttp` which owns the `NetworkConnection<StacksHttp>`, asked for less).  Will fix.